deepchecks/vision/dataset.py

```diff
@@ -211,3 +235,83 @@ def validate_dataset(cls, obj) -> 'VisionDataset':
                                        f'{type(obj).__name__}')
 
         return obj
+
+
+class InMemoryDataset(Dataset):
+    """Dataset implementation that gets all the data as in-memory list."""
+
+    def __init__(self, data: List):
+        self._data = data
+
+    def __len__(self):
+        return len(self._data)
+
+    def __getitem__(self, item):
+        return self._data[item]
+
+
+class FixedSampler(Sampler):
+    """Sampler which returns indices in a shuffled constant order."""
+
+    _length: int
+    _seed: int
+    _indices = None
+
+    def __init__(self, length: int, seed: int = 0, sample_size: int = None) -> None:
+        super().__init__(None)
+        self._length = length
+        self._seed = seed
+        if sample_size:
+            sample_size = min(sample_size, length)
+            np.random.seed(self._seed)
+            self._indices = np.random.choice(self._length, size=(sample_size,), replace=False)
+
+    def __iter__(self) -> Iterator[int]:
+        if self._indices:
+            for i in self._indices:
+                yield i
+        else:
+            for i in torch.randperm(self._length, generator=torch.Generator.manual_seed(self._seed)):
+                yield i
+
+    def __len__(self) -> int:
+        return len(self._indices) if self._indices else self._length
+
+
+def create_sample_loader(data_loader: DataLoader, sample_size: int, seed: int):
+    """Create a data loader with only a subset of the data."""
+    common_props_to_copy = {
+        'num_workers': data_loader.num_workers,
+        'collate_fn': data_loader.collate_fn,
+        'pin_memory': data_loader.pin_memory,
+        'timeout': data_loader.timeout,
+        'worker_init_fn': data_loader.worker_init_fn,
+        'prefetch_factor': data_loader.prefetch_factor,
+        'persistent_workers': data_loader.persistent_workers
+    }
+
+    generator = lambda: torch.Generator().manual_seed(seed)
+
+    dataset = data_loader.dataset
+    # IterableDataset doesn't work with samplers, so instead we manually copy all samples to memory and create
+    # new dataset that will contain them.
+    if isinstance(dataset, torch.utils.data.IterableDataset):
+        iter_length = 0
+        for _ in dataset:
+            iter_length += 1
+        sample_size = min(sample_size, iter_length)
+        np.random.seed(seed)
+        sample_indices = set(np.random.choice(iter_length, size=(sample_size,), replace=False))
```

Comment:
Isn't it an anti-pattern to iterate an iterator to find its length?
I understand we can't sample in a uniform distribution otherwise, but maybe it's better to limit the number of samples we are iterating on (and it can be a large number, e.g. 100000 or something). then, we can sample using [Reservoir sampling](https://en.wikipedia.org/wiki/Reservoir_sampling).

I know it might be an overkill solution now, but I'm afraid from use-cases where the data will be enormous..

Reply:
@noamzbr WDYT?

Reply:
Okay done that. Added a note in VisionDataset which is pretty important. might be good to put it somewhere else also.